From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Henrie <alexhenrie24@gmail.com>
Date: Sun, 3 Dec 2023 22:52:09 -0700
Subject: [PATCH] xdgmime: Handle realloc moving the buffer in
 _xdg_mime_(cache_)mime_type_subclass

Backport a patch from xdgmime missing from GLib.

For: https://gitlab.freedesktop.org/xdg/xdgmime/-/issues/37
---
 gio/xdgmime/xdgmime.c      | 26 ++++++++++++++------------
 gio/xdgmime/xdgmime.h      |  2 +-
 gio/xdgmime/xdgmimecache.c | 25 ++++++++++++++-----------
 gio/xdgmime/xdgmimecache.h |  2 +-
 4 files changed, 30 insertions(+), 25 deletions(-)

diff --git a/gio/xdgmime/xdgmime.c b/gio/xdgmime/xdgmime.c
index dce938a6f50d..a250d66a5a02 100644
--- a/gio/xdgmime/xdgmime.c
+++ b/gio/xdgmime/xdgmime.c
@@ -836,12 +836,12 @@ xdg_mime_is_super_type (const char *mime)
 int
 _xdg_mime_mime_type_subclass (const char *mime,
 			      const char *base,
-			      const char **seen)
+			      const char ***seen)
 {
   const char *umime, *ubase, *parent;
-  const char **parents;
+  const char **parents, **first_seen = NULL, **new_seen;
 
-  int first_seen = 0, i, ret = 0;
+  int i, ret = 0;
 
   if (_caches)
     return _xdg_mime_cache_mime_type_subclass (mime, base, NULL);
@@ -870,36 +870,38 @@ _xdg_mime_mime_type_subclass (const char *mime,
 
   if (!seen)
     {
-      seen = calloc (1, sizeof (char *));
-      first_seen = 1;
+      first_seen = calloc (1, sizeof (char *));
+      seen = &first_seen;
     }
 
   parents = _xdg_mime_parent_list_lookup (parent_list, umime);
   for (; parents && *parents; parents++)
     {
       parent = *parents;
 
       /* Detect and avoid buggy circular relationships */
-      for (i = 0; seen[i] != NULL; i++)
-        if (parent == seen[i])
+      for (i = 0; (*seen)[i] != NULL; i++)
+        if (parent == (*seen)[i])
           goto next_parent;
-      seen = realloc (seen, (i + 2) * sizeof (char *));
-      seen[i] = parent;
-      seen[i + 1] = NULL;
+      new_seen = realloc (*seen, (i + 2) * sizeof (char *));
+      if (!new_seen)
+        goto done;
+      new_seen[i] = parent;
+      new_seen[i + 1] = NULL;
+      *seen = new_seen;
 
       if (_xdg_mime_mime_type_subclass (parent, ubase, seen))
         {
           ret = 1;
           goto done;
         }
 
     next_parent:
       continue;
     }
 
 done:
-  if (first_seen)
-    free (seen);
+  free (first_seen);
   return ret;
 }
 
diff --git a/gio/xdgmime/xdgmime.h b/gio/xdgmime/xdgmime.h
index bbae1be79008..dc47339df65e 100644
--- a/gio/xdgmime/xdgmime.h
+++ b/gio/xdgmime/xdgmime.h
@@ -126,7 +126,7 @@ int          _xdg_mime_mime_type_equal             (const char *mime_a,
 						    const char *mime_b);
 int          _xdg_mime_mime_type_subclass          (const char *mime,
 						    const char *base,
-						    const char **seen);
+						    const char ***seen);
 const char  *_xdg_mime_unalias_mime_type           (const char *mime);
 
 
diff --git a/gio/xdgmime/xdgmimecache.c b/gio/xdgmime/xdgmimecache.c
index 9ea10f901dcc..affe68bdd6b2 100644
--- a/gio/xdgmime/xdgmimecache.c
+++ b/gio/xdgmime/xdgmimecache.c
@@ -902,12 +902,13 @@ is_super_type (const char *mime)
 int
 _xdg_mime_cache_mime_type_subclass (const char *mime,
 				    const char *base,
-				    const char **seen)
+				    const char ***seen)
 {
   const char *umime, *ubase, *parent;
+  const char **first_seen = NULL, **new_seen;
 
   xdg_uint32_t j;
-  int i, k, min, max, med, cmp, first_seen = 0, ret = 0;
+  int i, k, min, max, med, cmp, ret = 0;
 
   umime = _xdg_mime_cache_unalias_mime_type (mime);
   ubase = _xdg_mime_cache_unalias_mime_type (base);
@@ -936,8 +937,8 @@ _xdg_mime_cache_mime_type_subclass (const char *mime,
 
   if (!seen)
     {
-      seen = calloc (1, sizeof (char *));
-      first_seen = 1;
+      first_seen = calloc (1, sizeof (char *));
+      seen = &first_seen;
     }
 
   for (i = 0; _caches[i]; i++)
@@ -976,31 +977,33 @@ _xdg_mime_cache_mime_type_subclass (const char *mime,
 		  parent = cache->buffer + parent_offset;
 
 		  /* Detect and avoid buggy circular relationships */
-		  for (k = 0; seen[k] != NULL; k++)
-		    if (parent == seen[k])
+		  for (k = 0; (*seen)[k] != NULL; k++)
+		    if (parent == (*seen)[k])
 		      goto next_parent;
-		  seen = realloc (seen, (k + 2) * sizeof (char *));
-		  seen[k] = parent;
-		  seen[k + 1] = NULL;
+		  new_seen = realloc (*seen, (k + 2) * sizeof (char *));
+		  if (!new_seen)
+		    goto done;
+		  new_seen[k] = parent;
+		  new_seen[k + 1] = NULL;
+		  *seen = new_seen;
 
 		  if (_xdg_mime_cache_mime_type_subclass (parent, ubase, seen))
 		    {
 		      ret = 1;
 		      goto done;
 		    }
 
 		next_parent:
 		  continue;
 		}
 
 	      break;
 	    }
 	}
     }
 
 done:
-  if (first_seen)
-    free (seen);
+  free (first_seen);
   return ret;
 }
 
diff --git a/gio/xdgmime/xdgmimecache.h b/gio/xdgmime/xdgmimecache.h
index 0ac0b054fe28..315d86333f2b 100644
--- a/gio/xdgmime/xdgmimecache.h
+++ b/gio/xdgmime/xdgmimecache.h
@@ -71,7 +71,7 @@ int          _xdg_mime_cache_media_type_equal             (const char *mime_a,
 							   const char *mime_b);
 int          _xdg_mime_cache_mime_type_subclass           (const char *mime_a,
 							   const char *mime_b,
-							   const char **seen);
+							   const char ***seen);
 char       **_xdg_mime_cache_list_mime_parents		  (const char *mime);
 const char  *_xdg_mime_cache_unalias_mime_type            (const char *mime);
 int          _xdg_mime_cache_get_max_buffer_extents       (void);
